clients/src/test/java/org/apache/kafka/common/metrics/SensorTest.java

```diff
@@ -366,4 +368,60 @@ public void testUpdatingMetricConfigIsReflectedInTheSensor() {
 
         metrics.close();
     }
+
+    @Test
+    public void testConcurrentReadWriteAccessForMetrics() throws ExecutionException, InterruptedException {
+        final Metrics metrics = new Metrics(new MetricConfig().quota(Quota.upperBound(Double.MAX_VALUE))
+            .timeWindow(1, TimeUnit.MILLISECONDS)
+            .samples(100));
+        final Sensor sensor = metrics.sensor("sensor");
+
+        final int metricCount = 10000;
+        // Add a large number of metrics to increase the execution time of checkQuotas
+        for (int i = 0; i < metricCount; i++) {
+            sensor.add(metrics.metricName("test-metric" + i, "test-group" + i), new Rate());
+        }
+        sensor.record(10, 1, false);
+        CountDownLatch latch = new CountDownLatch(1);
+        ExecutorService executor = Executors.newFixedThreadPool(2);
+
+        // Use non-thread-safe methods for concurrent read and write of metrics.
+        Future<Throwable> worker = executor.submit(() -> {
+            try {
+                sensor.checkQuotasNonThreadSafe(3);
+                return null;
+            } catch (Throwable e) {
+                return e;
+            }
+        });
+        executor.submit(() -> {
+            sensor.add(metrics.metricName("test-metric-non-thread-safe", "test-group-non-thread-safe"), new Rate());
+            latch.countDown();
+        });
+
+        assertTrue(latch.await(5, TimeUnit.SECONDS), "If this failure happen frequently, we can try to increase the wait time");
+        assertTrue(worker.isDone());
+        assertNotNull(worker.get());
+        assertEquals(ConcurrentModificationException.class, worker.get().getClass());
```

Comment:
Will the test throw exceptions consistently? What if `sensor.checkQuotasNonThreadSafe(3)` ran before we are adding a new metric?

Reply:
This test will always throw an exception if new metrics are added during the process of iterating through the 'metrics'.

Reply:
@kamalcph has a valid point. We are relying on the assumption that iterating through the metrics takes a long time by setting metric count to 1000. In faster machines, this assumption might not hold. But I don't have a better solution to propose here.

Alternatively, to make it more simpler, we can structure the test as:

1. create two tasks 1 (checkQuotas) and 2 (add)
2. task 1 is submitted first but cannot complete until 2 is submitted.
3. after 2 is submitted both 1 and 2 will start executing.
4. 1 should take time (you can use mock such as https://github.com/apache/kafka/blob/b3ce2e54f40f2d1e287d8bfd196dc5dcdbd2046d/core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala#L378 to ensure that by adding thread.sleep). At the end of completion check if 2 has completed. It should not have.

To do this, you can create a latch called startFirst. Block the execution of first on it by doing something like this: https://github.com/apache/kafka/blob/b3ce2e54f40f2d1e287d8bfd196dc5dcdbd2046d/core/src/test/scala/unit/kafka/log/remote/RemoteIndexCacheTest.scala#L370 and then countdown this latch after 2 has been submitted. Also add a thread.pause for 100ms. 

With this, we can remove the additional method we have added just for testing and we will also be able to ensure that test fails with concurrent modification before and doesn't afterwards.

Reply:
@divijvaidya thank you very much for your explanation. However, I still have a question. Both 'add' and 'checkQuotas' are synchronized, so how can they be accessed simultaneously?